### PR TITLE
Alert manager listener

### PR DESCRIPTION
**Pull request by dkropachev**
https://trello.com/c/he5eCLA0/1397-alert-detector-to-detect-performance-issues-and-other-things

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [ ] ~~I added the relevant `backport` labels~~
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [x] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: sdcm/sct_events.py [hunk 1]
```diff
@@ -509,30 +538,6 @@ def run(self):
             except Exception:  # pylint: disable=broad-except
                 LOGGER.exception("Failed to write event to event.log")
 
-# This is an example of how we'll send info into Prometheus,
-# Currently it's not in use, since the data we want to show, doesn't fit Prometheus model,
-# we are using the GrafanaAnnotator
-
-
-class PrometheusDumper(threading.Thread):
```
Comment (bentsi) [inline sdcm/sct_events.py]: this is not relevant to your PR, please keep this example

File: sdcm/sct_events.py [hunk 2]
```diff
@@ -462,6 +462,35 @@ def __str__(self):
             super(SpotTerminationEvent, self).__str__(), self)
 
 
+class PrometheusAlertEvent(SctEvent):  # pylint: disable=too-many-instance-attributes
```
Comment (bentsi) [inline sdcm/sct_events.py]: `AlertManagerEvent` or `PrometheusAlertManagerEvent`

File: sdcm/sct_events.py [hunk 3]
```diff
@@ -462,6 +462,35 @@ def __str__(self):
...
+    def __init__(self, event_type:str, annotations: dict = None, startsAt=None, endsAt=None, updatedAt=None, fingerprint: str = None,
```
Comment (bentsi) [inline sdcm/sct_events.py]: this is not PEP 8, we use CamelCase only for class names

File: sdcm/sct_events.py (hunk 1 above)
Comment (bentsi) [reply sdcm/sct_events.py]: oh, you moved it to prometheus.py. never mind

File: sdcm/cluster.py [hunk 4]
```diff
@@ -915,6 +920,8 @@ def start_task_threads(self):
             self.start_journal_thread()
             self.start_backtrace_thread()
             self.start_db_log_reader_thread()
+        elif 'monitor' in self.name:
+            self.start_monitor_alert_manager()
```
Comment (bentsi) [inline sdcm/cluster.py]: `start_alert_manager_thread` - convention is that if something is starting a thread it should have a `thread` suffix as you see in above threads

File: sdcm/prometheus.py [hunk 5]
```diff
@@ -103,3 +106,103 @@ def event_stop(self, disrupt):
...
+
+
+class PrometheusAlertManagerListener:
+    interval = 10
+
+    def __init__(self, ip, port=9093):
+        self._alert_manager_url = f"http://{ip}:{port}/api/v2"
+
+    def start(self):
+        self.to_stop_flag = threading.Event()
```
Comment (bentsi) [inline sdcm/prometheus.py]: rule: never define instance attributes outside the constructor. (User of the class should understand what are the instance attributes from the constructor and not go over all the method definitions)

File: sdcm/sct_events.py (hunk 3 above)
Comment (dkropachev) [reply sdcm/sct_events.py]: problem that code fill these variables from the data it gets from openapi endpoint.
It is done to not have excess mapping code.

Comment (bentsi) [reply sdcm/sct_events.py]: `PrometheusAlert` will get the raw alert dict and then map the values to the instance attribute
```
class PrometheusAlert:
	def __init__(self, raw_alert):
		self.updated_at = raw_alert["updatedAt"]
		self.generator_url = raw_alert["generatorURL"]
                ...
```

File: sdcm/prometheus.py [hunk 6]
```diff
@@ -103,3 +106,103 @@ def event_stop(self, disrupt):
...
+            events_gauge.labels(event_type,  # pylint: disable=no-member
+                                getattr(message_data, 'type', ''),
+                                message_data.severity,
+                                getattr(message_data, 'node', '')).set(message_data.timestamp)
+
+    def terminate(self):
+        self.stop_event.set()
+
+
+class PrometheusAlert:
```
Comment (bentsi) [inline sdcm/prometheus.py]: As commented below, change this to 
```
class PrometheusAlert:
	def __init__(self, raw_alert):
		self.updated_at = raw_alert["updatedAt"]
		self.generator_url = raw_alert["generatorURL"]
                ...
                self._raw_alert = raw_alert
```

File: sdcm/sct_events.py [hunk 7]
```diff
@@ -462,6 +462,35 @@ def __str__(self):
...
+                 status: dict = None, generatorURL: str=None, labels: dict = None, severity=Severity.NORMAL, **kwargs):
```
Comment (bentsi) [inline sdcm/sct_events.py]: - I don't think we need the generatorURL in the Event, it doesn't have any valuable information
- if we have alert, the severity should be WARNING
- remove `**kwargs` (if there is something that needed, it should be passed in the constractor

File: sdcm/prometheus.py [hunk 8]
```diff
@@ -103,3 +106,103 @@ def event_stop(self, disrupt):
...
+    def __init__(self, annotations: dict = None, startsAt: str = None, endsAt: str = None, updatedAt: str = None,
+                 fingerprint: str = None, status: dict = None, generatorURL: str = None, labels: dict = None, **kwargs):
```
Comment (bentsi) [inline sdcm/prometheus.py]: remove `**kwargs`

File: sdcm/sct_events.py (hunk 7 above)
Comment (dkropachev) [reply sdcm/sct_events.py]: fingerprint - one alert rule can issue multiple alert instances with same name and starts_at, in this case fingnerprint is good thing to track down specific alert instance

File: sdcm/prometheus.py (hunk 5 above)
Comment (dkropachev) [reply sdcm/prometheus.py]: fixed.

File: sdcm/cluster.py (hunk 4 above)
Comment (dkropachev) [reply sdcm/cluster.py]: fixed

File: sdcm/sct_events.py (hunk 3 above)
Comment (dkropachev) [reply sdcm/sct_events.py]: fixed

File: sdcm/sct_events.py (hunk 2 above)
Comment (dkropachev) [reply sdcm/sct_events.py]: fixed

File: sdcm/prometheus.py (hunk 8 above)
Comment (dkropachev) [reply sdcm/prometheus.py]: fixed

File: sdcm/sct_events.py (hunk 7 above)
Comment (dkropachev) [reply sdcm/sct_events.py]: Other items are fixed

File: sdcm/prometheus.py (hunk 6 above)
Comment (dkropachev) [reply sdcm/prometheus.py]: fixed

**Comment by dkropachev** [comment]
Successfull run - https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dmitry/job/dmitry-byo-longevity/45/

File: unit_tests/test_prometheus.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    def test_alert_manager_listener_artificial_run(self):
+        with open('test_data/test_prometheus/test_alert_manager_listener_artificial_run.yaml') as f:
+            test_data = json.load(f)
+        i = PrometheusAlertManagerListenerArtificialTest(artificial_alerts=test_data['post'])
+        i.start()
+        result = i.get_result()
+        self.assertEqual(result, test_data['expected'])
+
+    @unittest.skip("TBD: dockerized prometheuse to be added")
+    def test_alert_manager_listener_real_run(self):
```
Comment (fruch) [inline unit_tests/test_prometheus.py]: :+1:  for manual kind of test, they really help next time we need to debug this part

**Pull request merged**